File: 120. Triangle/120. Triangle.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        https://discuss.leetcode.com/topic/19754/python-easy-to-understand-solutions-top-down-bottom-up
        """
        if not triangle:
            return 
        res = [[0 for i in xrange(len(row))] for row in triangle]
        res[0][0] = triangle[0][0]
        for i in xrange(1, len(triangle)):
            for j in xrange(len(triangle[i])):
                if not j:
                    res[i][j] = res[i - 1][j] + triangle[i][j]
                elif j == len(triangle[i]) - 1:
                    res[i][j] = res[i - 1][j - 1] + triangle[i][j]
                else:
                    res[i][j] = min(res[i - 1][j - 1], res[i - 1][j]) + triangle[i][j]
        return min(res[-1])

File: 120. Triangle/120. Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
    int** sumPath;
    int r, c, m = triangle.size();
    int min;
    int* triangleColSizes = (int*) malloc(sizeof(int*) * m);
    for (r = 0; r < m; r++)
        *(triangleColSizes + r) = triangle[r].size();
    sumPath = (int **) malloc(sizeof(int*) * m);
    for (r = 0; r < m; r++)
        sumPath[r] = (int*) malloc(sizeof(int) * *(triangleColSizes + r));
    for (r = 0; r < m; r++)
        for (c = 0; c < *(triangleColSizes + r); c++)
            sumPath[r][c] = 0;
    sumPath[0][0] = triangle[0][0];
    for (r = 1; r < m; r++)
        for (c = 0; c < *(triangleColSizes + r); c++){
            if (c == 0){
                sumPath[r][c] = sumPath[r - 1][c] + triangle[r][c];
            }else if (c == *(triangleColSizes + r) - 1){
                sumPath[r][c] = sumPath[r - 1][c - 1] + triangle[r][c];
            }
            else{
                if (sumPath[r - 1][c - 1] > sumPath[r - 1][c])
                    sumPath[r][c] = triangle[r][c] + sumPath[r - 1][c];
                else
                    sumPath[r][c] = triangle[r][c] + sumPath[r - 1][c - 1];
            }
        }
    min = sumPath[m - 1][0];
    for (c = 0; c < *(triangleColSizes + m - 1); c++)
        if (min > sumPath[m - 1][c])
            min = sumPath[m - 1][c];
    return min;
}
};
